### ISSUE TITLE
Refactor use generator for middleware decorator

### ISSUE THREAD

**Issue by andrewgryan** [informatics-lab/forest#226]
**TL;DR** The initial generator idea can be completely replaced by action generator functions and a simple loop in `Store.dispatch()`

### Initial idea on generator refactor

`@middleware` wraps a function/method 3 layers deep so the `Traceback` stack for multiple middlewares is `3N` deep. This makes it difficult to debug fairly trivial problems.

```python
def gen_middlware(f):
     @functools.wraps(f)
     def wrapper():
           store = yield
           next_dispatch = yield
           yield functools.partial(f, store, next_dispatch)
    return wrapper
```

Can be used as follows inside `Store` constructor to generate middleware call signature

```python
g.send(None)  # Needed to start generator
g.send(store)
f = g.send(next_dispatch)
```
`f` then has the call signature `f(action)` but with `store` and `next_dispatch` baked in as desired. Since there is no nesting the call stack is shallower. This solves the `3` in our `3N` call stack nesting.

To solve the chained nature of `next_dispatch(next_dispatch(next_dispatch(...)))` perhaps we could use a loop instead of recursion.

**Comment by andrewgryan** [comment]
Again I think generators can rescue us from deeply nested call stacks

```python
def some_middleware(store, next_dispatch, action):
     if action["kind"] == 'some_prop':
         yield next_dispatch(different_action())  # custom additional action
    yield next_dispatch(action)  # default behaviour
```
Then `Store.dispatch` loops through each middleware until all actions and additional actions have been passed through dispatchers

```python
for dispatch in self.dispatchers:
     for callable in dispatch(action):
          callable(action)
```

The `Traceback` should be just the call stack for the particular action taking place. Perhaps we lose the who called who aspect of the current implementation of `@middleware` and `Store.dispatch`

**Comment by andrewgryan** [comment]
In fact, generators that yield actions are all that we need.

```python
def some_middleware(store, action):
     if action["kind"] == 'some_prop':
         yield different_action()  # custom additional action
    yield action  # default behaviour
```

```python
next_actions  = []
for dispatch in self.dispatchers:
    if next_actions:
        for next_action in next_actions:
            dispatch(next_action)
    next_actions = dispatch(in_action)
```

Instead of chaining `dispatch` methods together we pass lists of actions to the next dispatch in the loop

**Comment by andrewgryan** [comment]
In Python middleware can be "action generators" given a `Store` and a current `action` they generate further actions to take, swallow actions by not yielding, or enhance actions by yielding the same action with extra data

**Comment by andrewgryan** [comment]
This also eases testing as `list(middleware_func(a))` returns a list of concrete actions. The call stack currying mechanism has been completely removed and we are left with logic and data

**Comment by andrewgryan** [comment]
#236 closed this issue
